### ISSUE TITLE
[Bug]: Primary PostgreSQL Instance Not Gracefully Shutting Down and Missing WAL File

### ISSUE THREAD

**Issue by orenzah** [cloudnative-pg/cloudnative-pg#3680]
### Is there an existing issue already for this bug?

- [X] I have searched for an existing issue, and could not find anything. I believe this is a new bug.

### I have read the troubleshooting guide

- [X] I have read the troubleshooting guide and I think this is a new bug.

### I am running a supported version of CloudNativePG

- [X] I have read the troubleshooting guide and I think this is a new bug.

### Contact Details

oren@aperio.ai

### Version

1.22.0

### What version of Kubernetes are you using?

1.27

### What is your Kubernetes environment?

Cloud: Amazon EKS

### How did you install the operator?

Helm

### What happened?

We deployed the operator using a Helm chart and the PostgreSQL cluster resource using a separate Helm chart. Three instances of PostgreSQL were deployed on EC2 Spot instances. After a few hours, it appears that the primary instance (postgres-1) did not gracefully shut down, leading to the absence of the relevant Write-Ahead Logging (WAL) file.

Subsequently, attempts to use pg_rewind have failed, with the process encountering an empty file. Additionally, the file "pg_xact/0000" is also empty, contributing to the inability to resolve the issue and leaving the old primary instance in a stuck state.

### Cluster resource

```shell
apiVersion: postgresql.cnpg.io/v1
kind: Cluster
metadata:
  annotations:
    meta.helm.sh/release-name: postgres-cn
    meta.helm.sh/release-namespace: stg
  creationTimestamp: "2024-01-09T12:06:49Z"
  generation: 4
  labels:
    app.kubernetes.io/managed-by: Helm
  name: postgres
  namespace: stg
  resourceVersion: "346234286"
  uid: 69ddea55-1c72-4e9f-abf6-aa2e43a32db4
spec:
  affinity:
    podAntiAffinityType: preferred
  bootstrap:
    initdb:
      database: app
      encoding: UTF8
      import:
        databases:
        - webapi
        source:
          externalCluster: xxxxxx
        type: microservice
      localeCType: C
      localeCollate: C
      owner: app
  enableSuperuserAccess: true
  externalClusters:
  - connectionParameters:
      dbname: webapi
      host: xxxxxxxxxxxx
      user: postgres
    name: xxxxxxx
    password:
      key: postgres-password
      name: xxxxxxxxxxxxxx
  failoverDelay: 0
  imageName: xxxxxxx/postgresql:16.1-xxxxxxxxx
  inheritedMetadata:
    labels:
      app.kubernetes.io/name: postgres
      app.kubernetes.io/version: "16.1"
  instances: 3
  logLevel: info
  maxSyncReplicas: 2
  minSyncReplicas: 1
  monitoring:
    customQueriesConfigMap:
    - key: queries
      name: cnpg-default-monitoring
    disableDefaultQueries: false
    enablePodMonitor: false
  postgresGID: 26
  postgresUID: 26
  postgresql:
    parameters:
      archive_mode: "on"
      archive_timeout: 5min
      dynamic_shared_memory_type: posix
      log_destination: csvlog
      log_directory: /controller/log
      log_filename: postgres
      log_rotation_age: "0"
      log_rotation_size: "0"
      log_truncate_on_rotation: "false"
      logging_collector: "on"
      max_parallel_workers: "32"
      max_replication_slots: "32"
      max_worker_processes: "32"
      shared_memory_type: mmap
      shared_preload_libraries: ""
      ssl_max_protocol_version: TLSv1.3
      ssl_min_protocol_version: TLSv1.3
      wal_keep_size: 512MB
      wal_receiver_timeout: 5s
      wal_sender_timeout: 5s
    syncReplicaElectionConstraint:
      enabled: false
  primaryUpdateMethod: restart
  primaryUpdateStrategy: unsupervised
  replicationSlots:
    highAvailability:
      enabled: true
      slotPrefix: _cnpg_
    updateInterval: 30
  resources: {}
  smartShutdownTimeout: 180
  startDelay: 3600
  stopDelay: 1800
  storage:
    resizeInUseVolumes: true
    size: 30Gi
    storageClass: gp3
  superuserSecret:
    name: postgres-superuser
  switchoverDelay: 3600
status:
  certificates:
    clientCASecret: postgres-ca
    expirations:
      postgres-ca: 2024-04-08 12:01:49 +0000 UTC
      postgres-replication: 2024-04-08 12:01:49 +0000 UTC
      postgres-server: 2024-04-08 12:01:49 +0000 UTC
    replicationTLSSecret: postgres-replication
    serverAltDNSNames:
    - postgres-rw
    - postgres-rw.stg
    - postgres-rw.stg.svc
    - postgres-r
    - postgres-r.stg
    - postgres-r.stg.svc
    - postgres-ro
    - postgres-ro.stg
    - postgres-ro.stg.svc
    serverCASecret: postgres-ca
    serverTLSSecret: postgres-server
  cloudNativePGCommitHash: 86b9dc80
  cloudNativePGOperatorHash: 411a67468c5e1357d387221ce5b11d906bd33d62324846a8c776d206b6c94162
  conditions:
  - lastTransitionTime: "2024-01-10T08:20:16Z"
    message: Cluster Is Not Ready
    reason: ClusterIsNotReady
    status: "False"
    type: Ready
  - lastTransitionTime: "2024-01-10T12:18:14Z"
    message: Continuous archiving is working
    reason: ContinuousArchivingSuccess
    status: "True"
    type: ContinuousArchiving
  configMapResourceVersion:
    metrics:
      cnpg-default-monitoring: "345484626"
  currentPrimary: postgres-2
  currentPrimaryTimestamp: "2024-01-10T08:21:59.663283Z"
  healthyPVC:
  - postgres-1
  - postgres-2
  - postgres-3
  instanceNames:
  - postgres-1
  - postgres-2
  - postgres-3
  instances: 3
  instancesReportedState:
    postgres-1:
      isPrimary: false
    postgres-2:
      isPrimary: true
      timeLineID: 2
    postgres-3:
      isPrimary: false
      timeLineID: 2
  instancesStatus:
    healthy:
    - postgres-2
    - postgres-3
    replicating:
    - postgres-1
  latestGeneratedNode: 3
  managedRolesStatus: {}
  phase: Waiting for the instances to become active
  phaseReason: Some instances are not yet active. Please wait.
  poolerIntegrations:
    pgBouncerIntegration: {}
  pvcCount: 3
  readService: postgres-r
  readyInstances: 2
  secretsResourceVersion:
    applicationSecretVersion: "345484596"
    clientCaSecretVersion: "345484593"
    externalClusterSecretVersion:
      stg-postgres-postgresql: "231319709"
    replicationSecretVersion: "345484595"
    serverCaSecretVersion: "345484593"
    serverSecretVersion: "345484594"
    superuserSecretVersion: "341642273"
  targetPrimary: postgres-2
  targetPrimaryTimestamp: "2024-01-10T08:20:16.701057Z"
  timelineID: 2
  topology:
    instances:
      postgres-1: {}
      postgres-2: {}
      postgres-3: {}
    nodesUsed: 2
    successfullyExtracted: true
  writeService: postgres-rw
```

### Relevant log output

```shell
{"level":"error","ts":"2024-01-10T10:58:19Z","msg":"Reconciler error","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"postgres","namespace":"stg"},"namespace":"stg","name":"postgres","reconcileID":"9b64cee9-7e71-4e05-9bb0-e1cbaee76c9d","error":"while activating instance: error starting PostgreSQL instance: exit status 1","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\tpkg/mod/sigs.k8s.io/controller-runtime@v0.16.3/pkg/internal/controller/controller.go:329\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\tpkg/mod/sigs.k8s.io/controller-runtime@v0.16.3/pkg/internal/controller/controller.go:266\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\tpkg/mod/sigs.k8s.io/controller-runtime@v0.16.3/pkg/internal/controller/controller.go:227"}
{"level":"info","ts":"2024-01-10T10:58:28Z","msg":"Cluster status","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"postgres","namespace":"stg"},"namespace":"stg","name":"postgres","reconcileID":"b6dbc5ac-aab1-4528-bc0c-f67aa8dac9b1","uuid":"2f587c83-afa7-11ee-9aef-929973a308f2","logging_pod":"postgres-1","currentPrimary":"postgres-2","targetPrimary":"postgres-2"}
{"level":"info","ts":"2024-01-10T10:58:28Z","msg":"This is an old primary instance, waiting for the switchover to finish","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"postgres","namespace":"stg"},"namespace":"stg","name":"postgres","reconcileID":"b6dbc5ac-aab1-4528-bc0c-f67aa8dac9b1","uuid":"2f587c83-afa7-11ee-9aef-929973a308f2","logging_pod":"postgres-1","currentPrimary":"postgres-2","targetPrimary":"postgres-2"}
{"level":"info","ts":"2024-01-10T10:58:28Z","msg":"Switchover completed","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"postgres","namespace":"stg"},"namespace":"stg","name":"postgres","reconcileID":"b6dbc5ac-aab1-4528-bc0c-f67aa8dac9b1","uuid":"2f587c83-afa7-11ee-9aef-929973a308f2","logging_pod":"postgres-1","targetPrimary":"postgres-2","currentPrimary":"postgres-2"}
{"level":"info","ts":"2024-01-10T10:58:28Z","msg":"Waiting for the new primary to be available","logging_pod":"postgres-1","primaryConnInfo":"host=postgres-rw user=streaming_replica port=5432 sslkey=/controller/certificates/streaming_replica.key sslcert=/controller/certificates/streaming_replica.crt sslrootcert=/controller/certificates/server-ca.crt application_name=postgres-1 sslmode=verify-ca dbname=postgres connect_timeout=5"}
{"level":"info","ts":"2024-01-10T10:58:28Z","msg":"Extracting pg_controldata information","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"postgres","namespace":"stg"},"namespace":"stg","name":"postgres","reconcileID":"b6dbc5ac-aab1-4528-bc0c-f67aa8dac9b1","uuid":"2f587c83-afa7-11ee-9aef-929973a308f2","logging_pod":"postgres-1","reason":"before pg_rewind"}
{"level":"info","ts":"2024-01-10T10:58:28Z","logger":"pg_controldata","msg":"pg_control version number:            1300\nCatalog version number:               202307071\nDatabase system identifier:           7322068936411099156\nDatabase cluster state:               shut down\npg_control last modified:             Wed 10 Jan 2024 08:19:35 AM UTC\nLatest checkpoint location:           2/AA000028\nLatest checkpoint's REDO location:    2/AA000028\nLatest checkpoint's REDO WAL file:    0000000100000002000000AA\nLatest checkpoint's TimeLineID:       1\nLatest checkpoint's PrevTimeLineID:   1\nLatest checkpoint's full_page_writes: on\nLatest checkpoint's NextXID:          0:36923\nLatest checkpoint's NextOID:          25092\nLatest checkpoint's NextMultiXactId:  1\nLatest checkpoint's NextMultiOffset:  0\nLatest checkpoint's oldestXID:        722\nLatest checkpoint's oldestXID's DB:   1\nLatest checkpoint's oldestActiveXID:  0\nLatest checkpoint's oldestMultiXid:   1\nLatest checkpoint's oldestMulti's DB: 1\nLatest checkpoint's oldestCommitTsXid:0\nLatest checkpoint's newestCommitTsXid:0\nTime of latest checkpoint:            Wed 10 Jan 2024 08:19:35 AM UTC\nFake LSN counter for unlogged rels:   0/3E8\nMinimum recovery ending location:     0/0\nMin recovery ending loc's timeline:   0\nBackup start location:                0/0\nBackup end location:                  0/0\nEnd-of-backup record required:        no\nwal_level setting:                    logical\nwal_log_hints setting:                on\nmax_connections setting:              100\nmax_worker_processes setting:         32\nmax_wal_senders setting:              10\nmax_prepared_xacts setting:           0\nmax_locks_per_xact setting:           64\ntrack_commit_timestamp setting:       off\nMaximum data alignment:               8\nDatabase block size:                  8192\nBlocks per segment of large relation: 131072\nWAL block size:                       8192\nBytes per WAL segment:                16777216\nMaximum length of identifiers:        64\nMaximum columns in an index:          32\nMaximum size of a TOAST chunk:        1996\nSize of a large-object chunk:         2048\nDate/time type storage:               64-bit integers\nFloat8 argument passing:              by value\nData page checksum version:           0\nMock authentication nonce:            4e9ac5de19e20369e90184039bdcf5c867fa08bd811982f2edb29d1f11d8fd7b\n","pipe":"stdout","logging_pod":"postgres-1"}
{"level":"info","ts":"2024-01-10T10:58:28Z","msg":"Starting up pg_rewind","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"postgres","namespace":"stg"},"namespace":"stg","name":"postgres","reconcileID":"b6dbc5ac-aab1-4528-bc0c-f67aa8dac9b1","uuid":"2f587c83-afa7-11ee-9aef-929973a308f2","logging_pod":"postgres-1","pgdata":"/var/lib/postgresql/data/pgdata","options":["-P","--source-server","host=postgres-rw user=streaming_replica port=5432 sslkey=/controller/certificates/streaming_replica.key sslcert=/controller/certificates/streaming_replica.crt sslrootcert=/controller/certificates/server-ca.crt application_name=postgres-1 sslmode=verify-ca dbname=postgres","--target-pgdata","/var/lib/postgresql/data/pgdata","--restore-target-wal"]}
{"level":"info","ts":"2024-01-10T10:58:28Z","logger":"pg_rewind","msg":"pg_rewind: connected to server","pipe":"stderr","logging_pod":"postgres-1"}
{"level":"info","ts":"2024-01-10T10:58:28Z","logger":"pg_rewind","msg":"pg_rewind: servers diverged at WAL location 2/A9000860 on timeline 1","pipe":"stderr","logging_pod":"postgres-1"}
{"level":"info","ts":"2024-01-10T10:58:28Z","logger":"pg_rewind","msg":"pg_rewind: error: could not read file \"/var/lib/postgresql/data/pgdata/pg_wal/0000000100000002000000A8\": read 0 of 8192","pipe":"stderr","logging_pod":"postgres-1"}
{"level":"info","ts":"2024-01-10T10:58:28Z","logger":"pg_rewind","msg":"pg_rewind: error: could not find previous WAL record at 2/A8241C48","pipe":"stderr","logging_pod":"postgres-1"}
{"level":"error","ts":"2024-01-10T10:58:28Z","msg":"Failed to execute pg_rewind","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"postgres","namespace":"stg"},"namespace":"stg","name":"postgres","reconcileID":"b6dbc5ac-aab1-4528-bc0c-f67aa8dac9b1","uuid":"2f587c83-afa7-11ee-9aef-929973a308f2","logging_pod":"postgres-1","options":["-P","--source-server","host=postgres-rw user=streaming_replica port=5432 sslkey=/controller/certificates/streaming_replica.key sslcert=/controller/certificates/streaming_replica.crt sslrootcert=/controller/certificates/server-ca.crt application_name=postgres-1 sslmode=verify-ca dbname=postgres","--target-pgdata","/var/lib/postgresql/data/pgdata","--restore-target-wal"],"error":"exit status 1","stacktrace":"github.com/cloudnative-pg/cloudnative-pg/pkg/management/log.(*logger).Error\n\tpkg/management/log/log.go:128\ngithub.com/cloudnative-pg/cloudnative-pg/pkg/management/postgres.(*Instance).Rewind\n\tpkg/management/postgres/instance.go:972\ngithub.com/cloudnative-pg/cloudnative-pg/internal/management/controller.(*InstanceReconciler).verifyPgDataCoherenceForPrimary\n\tinternal/management/controller/instance_startup.go:235\ngithub.com/cloudnative-pg/cloudnative-pg/internal/management/controller.(*InstanceReconciler).initialize\n\tinternal/management/controller/instance_controller.go:342\ngithub.com/cloudnative-pg/cloudnative-pg/internal/management/controller.(*InstanceReconciler).Reconcile\n\tinternal/management/controller/instance_controller.go:141\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile\n\tpkg/mod/sigs.k8s.io/controller-runtime@v0.16.3/pkg/internal/controller/controller.go:119\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\tpkg/mod/sigs.k8s.io/controller-runtime@v0.16.3/pkg/internal/controller/controller.go:316\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\tpkg/mod/sigs.k8s.io/controller-runtime@v0.16.3/pkg/internal/controller/controller.go:266\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\tpkg/mod/sigs.k8s.io/controller-runtime@v0.16.3/pkg/internal/controller/controller.go:227"}
{"level":"info","ts":"2024-01-10T10:58:28Z","msg":"pg_rewind failed, starting the server to complete the crash recovery","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"postgres","namespace":"stg"},"namespace":"stg","name":"postgres","reconcileID":"b6dbc5ac-aab1-4528-bc0c-f67aa8dac9b1","uuid":"2f587c83-afa7-11ee-9aef-929973a308f2","logging_pod":"postgres-1","err":"error executing pg_rewind: exit status 1"}
{"level":"info","ts":"2024-01-10T10:58:28Z","msg":"Waiting for server to complete crash recovery","logging_pod":"postgres-1"}
{"level":"info","ts":"2024-01-10T10:58:28Z","msg":"Starting up instance","logging_pod":"postgres-1","pgdata":"/var/lib/postgresql/data/pgdata","options":["start","-w","-D","/var/lib/postgresql/data/pgdata","-o","-c port=5432 -c unix_socket_directories=/controller/run","-t 40000000"]}
{"level":"info","ts":"2024-01-10T10:58:28Z","logger":"pg_ctl","msg":"waiting for server to start....2024-01-10 10:58:28.483 UTC [348] LOG:  redirecting log output to logging collector process","pipe":"stdout","logging_pod":"postgres-1"}
{"level":"info","ts":"2024-01-10T10:58:28Z","logger":"pg_ctl","msg":"2024-01-10 10:58:28.483 UTC [348] HINT:  Future log output will appear in directory \"/controller/log\".","pipe":"stdout","logging_pod":"postgres-1"}
{"level":"info","ts":"2024-01-10T10:58:28Z","logger":"postgres","msg":"2024-01-10 10:58:28.484 UTC [348] LOG:  ending log output to stderr","source":"/controller/log/postgres","logging_pod":"postgres-1"}
{"level":"info","ts":"2024-01-10T10:58:28Z","logger":"postgres","msg":"2024-01-10 10:58:28.484 UTC [348] HINT:  Future log output will go to log destination \"csvlog\".","source":"/controller/log/postgres","logging_pod":"postgres-1"}
{"level":"info","ts":"2024-01-10T10:58:28Z","logger":"postgres","msg":"record","logging_pod":"postgres-1","record":{"log_time":"2024-01-10 10:58:28.484 UTC","process_id":"348","session_id":"659e7854.15c","session_line_num":"1","session_start_time":"2024-01-10 10:58:28 UTC","transaction_id":"0","error_severity":"LOG","sql_state_code":"00000","message":"ending log output to stderr","hint":"Future log output will go to log destination \"csvlog\".","backend_type":"postmaster","query_id":"0"}}
{"level":"info","ts":"2024-01-10T10:58:28Z","logger":"postgres","msg":"record","logging_pod":"postgres-1","record":{"log_time":"2024-01-10 10:58:28.484 UTC","process_id":"348","session_id":"659e7854.15c","session_line_num":"2","session_start_time":"2024-01-10 10:58:28 UTC","transaction_id":"0","error_severity":"LOG","sql_state_code":"00000","message":"starting PostgreSQL 16.1 (Debian 16.1-1.pgdg110+1) on x86_64-pc-linux-gnu, compiled by gcc (Debian 10.2.1-6) 10.2.1 20210110, 64-bit","backend_type":"postmaster","query_id":"0"}}
{"level":"info","ts":"2024-01-10T10:58:28Z","logger":"postgres","msg":"record","logging_pod":"postgres-1","record":{"log_time":"2024-01-10 10:58:28.484 UTC","process_id":"348","session_id":"659e7854.15c","session_line_num":"3","session_start_time":"2024-01-10 10:58:28 UTC","transaction_id":"0","error_severity":"LOG","sql_state_code":"00000","message":"listening on IPv4 address \"0.0.0.0\", port 5432","backend_type":"postmaster","query_id":"0"}}
{"level":"info","ts":"2024-01-10T10:58:28Z","logger":"postgres","msg":"record","logging_pod":"postgres-1","record":{"log_time":"2024-01-10 10:58:28.484 UTC","process_id":"348","session_id":"659e7854.15c","session_line_num":"4","session_start_time":"2024-01-10 10:58:28 UTC","transaction_id":"0","error_severity":"LOG","sql_state_code":"00000","message":"listening on IPv6 address \"::\", port 5432","backend_type":"postmaster","query_id":"0"}}
{"level":"info","ts":"2024-01-10T10:58:28Z","logger":"postgres","msg":"record","logging_pod":"postgres-1","record":{"log_time":"2024-01-10 10:58:28.488 UTC","process_id":"348","session_id":"659e7854.15c","session_line_num":"5","session_start_time":"2024-01-10 10:58:28 UTC","transaction_id":"0","error_severity":"LOG","sql_state_code":"00000","message":"listening on Unix socket \"/controller/run/.s.PGSQL.5432\"","backend_type":"postmaster","query_id":"0"}}
{"level":"info","ts":"2024-01-10T10:58:28Z","logger":"postgres","msg":"record","logging_pod":"postgres-1","record":{"log_time":"2024-01-10 10:58:28.495 UTC","process_id":"352","session_id":"659e7854.160","session_line_num":"1","session_start_time":"2024-01-10 10:58:28 UTC","transaction_id":"0","error_severity":"LOG","sql_state_code":"00000","message":"database system was shut down at 2024-01-10 08:19:35 UTC","backend_type":"startup","query_id":"0"}}
{"level":"info","ts":"2024-01-10T10:58:28Z","logger":"postgres","msg":"record","logging_pod":"postgres-1","record":{"log_time":"2024-01-10 10:58:28.495 UTC","process_id":"352","session_id":"659e7854.160","session_line_num":"2","session_start_time":"2024-01-10 10:58:28 UTC","transaction_id":"0","error_severity":"FATAL","sql_state_code":"XX000","message":"could not access status of transaction 36923","detail":"Could not read from file \"pg_xact/0000\" at offset 8192: read too few bytes.","backend_type":"startup","query_id":"0"}}
{"level":"info","ts":"2024-01-10T10:58:28Z","logger":"postgres","msg":"record","logging_pod":"postgres-1","record":{"log_time":"2024-01-10 10:58:28.496 UTC","process_id":"348","session_id":"659e7854.15c","session_line_num":"6","session_start_time":"2024-01-10 10:58:28 UTC","transaction_id":"0","error_severity":"LOG","sql_state_code":"00000","message":"startup process (PID 352) exited with exit code 1","backend_type":"postmaster","query_id":"0"}}
{"level":"info","ts":"2024-01-10T10:58:28Z","logger":"postgres","msg":"record","logging_pod":"postgres-1","record":{"log_time":"2024-01-10 10:58:28.497 UTC","process_id":"348","session_id":"659e7854.15c","session_line_num":"7","session_start_time":"2024-01-10 10:58:28 UTC","transaction_id":"0","error_severity":"LOG","sql_state_code":"00000","message":"aborting startup due to startup process failure","backend_type":"postmaster","query_id":"0"}}
{"level":"info","ts":"2024-01-10T10:58:28Z","logger":"postgres","msg":"record","logging_pod":"postgres-1","record":{"log_time":"2024-01-10 10:58:28.497 UTC","process_id":"348","session_id":"659e7854.15c","session_line_num":"8","session_start_time":"2024-01-10 10:58:28 UTC","transaction_id":"0","error_severity":"LOG","sql_state_code":"00000","message":"database system is shut down","backend_type":"postmaster","query_id":"0"}}
{"level":"info","ts":"2024-01-10T10:58:28Z","logger":"pg_ctl","msg":" stopped waiting","pipe":"stdout","logging_pod":"postgres-1"}
{"level":"info","ts":"2024-01-10T10:58:28Z","logger":"pg_ctl","msg":"pg_ctl: could not start server","pipe":"stderr","logging_pod":"postgres-1"}
{"level":"info","ts":"2024-01-10T10:58:28Z","logger":"pg_ctl","msg":"Examine the log output.","pipe":"stderr","logging_pod":"postgres-1"}
{"level":"info","ts":"2024-01-10T10:58:28Z","msg":"Exited log pipe","fileName":"/controller/log/postgres.csv","logging_pod":"postgres-1"}
{"level":"error","ts":"2024-01-10T10:58:28Z","msg":"Reconciler error","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"postgres","namespace":"stg"},"namespace":"stg","name":"postgres","reconcileID":"b6dbc5ac-aab1-4528-bc0c-f67aa8dac9b1","error":"while activating instance: error starting PostgreSQL instance: exit status 1","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\tpkg/mod/sigs.k8s.io/controller-runtime@v0.16.3/pkg/internal/controller/controller.go:329\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\tpkg/mod/sigs.k8s.io/controller-runtime@v0.16.3/pkg/internal/controller/controller.go:266\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\tpkg/mod/sigs.k8s.io/controller-runtime@v0.16.3/pkg/internal/controller/controller.go:227"}
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by gbartolini** [comment]
From what I can see, it looks like an issue with the underlying file system (data corruption of that file). You need to recreate the PVC of that instance.

**Comment by orenzah** [comment]
It happens 2 days in a row, we are using EC2 spot instances, and we managed to solve it by deleting the problematic PVC (gp3)  but that manual PVC creation every day is probably not the desired behavior.

**Comment by gbartolini** [comment]
This might be related to #3698. Give us some time to investigate.

**Comment by janosmiko** [comment]
I face the same issue in multiple clusters in the same k8s cluster. I think it occurs when a network error happens.

The only way to resolve it is to delete the PVC of the failing pod and remove the pod. In this case, the operator is going to recreate the failing replica.

**Comment by ensdhvi** [comment]
We are facing the same problem on Google GKE clusters. After each node upgrade random instances have problems with recovering. The only workaround we found is removing one of failing instance PVC.

**Comment by movergan** [comment]
Just encountered into the same issue in GKE

**Comment by suleimi** [comment]
I keep running into this every other day

**Comment by maaft** [comment]
Also discovered this now.. is CNPG team aware of this? Whom to ping about this issue?

**Comment by oberai07** [comment]
I am facing similar issue. Do we have any fix available for below issue ? I tried to delete the pod but did not help.

`{"level":"error","ts":"2024-02-29T07:12:33Z","msg":"PostgreSQL process exited with errors","logging_pod":"harbor-postgres-3","error":"exit status 1","stacktrace":"github.com/cloudnative-pg/cloudnative-pg/pkg/management/log.(*logger).Error\n\tpkg/management/log/log.go:128\ngithub.com/cloudnative-pg/cloudnative-pg/internal/cmd/manager/instance/run/lifecycle.(*PostgresLifecycle).Start\n\tinternal/cmd/manager/instance/run/lifecycle/lifecycle.go:98\nsigs.k8s.io/controller-runtime/pkg/manager.(*runnableGroup).reconcile.func1\n\tpkg/mod/sigs.k8s.io/controller-runtime@v0.16.3/pkg/manager/runnable_group.go:223"}`

**Comment by janosmiko** [comment]
@gbartolini is it possible to modify the health check behavior of the pods managed by CNPG?

**Comment by suleimi** [comment]
> This might be related to #3698. Give us some time to investigate.

@gbartolini Any updates on this would be greatly appreciated 🙏🏿

**Comment by brettinternet** [comment]
The only way I've been able to recover from this is to delete both the PVC and Pod

**Comment by Dev-Destructor** [comment]
Same Issue here. Luckily, Our production cluster survived because one of the 3 replicas was working. Deleting both PVC and Pod solved the issue but it is not a permanent solution. 

we use GKE

**Comment by maximemoreillon** [comment]
I'm facing the same issue using the CNPG community addon for Microk8s.
After a reboot of the cluster, pods sometimes end up in `0/1 Running` with the `Reconciler error`

**Comment by maximemoreillon** [comment]
Indeed, it seems good for my cluster too

**Comment by cybercoder** [comment]
I updated the operator to the latest version yesterday and restored the database; today it's the same again. It makes me nervous and angry

**Comment by amalic** [comment]
I think I am facing the similar issue. I am running CNPG on AWS Spot Instances provisioned through Karpenter Cluster Autoscaler. Currently I have only one of the three instances working. The other two instance don't seem to come up. 

I am using CNPG Operator  **v1.23.1**

Seeing error messages like 
```JSON
{"level":"info","ts":"2024-08-06T10:12:20Z","logger":"postgres","msg":"record","logging_pod":"pg-cluster-3","record":{"log_time":"2024-08-06 10:12:20.737 UTC","process_id":"18434","session_id":"66b1f704.4802","session_line_num":"2","session_start_time":"2024-08-06 10:12:20 UTC","transaction_id":"0","error_severity":"FATAL","sql_state_code":"XX000","message":"could not receive data from WAL stream: ERROR:  requested WAL segment 0000000A000001BB00000044 has already been removed","backend_type":"walreceiver"}}
{"level":"info","ts":"2024-08-06T10:12:20Z","logger":"postgres","msg":"record","logging_pod":"pg-cluster-3","record":{"log_time":"2024-08-06 10:12:20.913 UTC","user_name":"postgres","database_name":"postgres","process_id":"18441","connection_from":"[local]","session_id":"66b1f704.4809","session_line_num":"1","session_start_time":"2024-08-06 10:12:20 UTC","transaction_id":"0","error_severity":"FATAL","sql_state_code":"57P03","message":"the database system is starting up","backend_type":"client backend"}}
```

and
```JSON
{"level":"info","ts":"2024-08-06T10:12:21Z","msg":"DB not available, will retry","controller":"cluster","controllerGroup":"postgresql.cnpg.io","controllerKind":"Cluster","Cluster":{"name":"pg-cluster","namespace":"xxx-development"},"namespace":"xxx-development","name":"pg-cluster","reconcileID":"f6812cba-7a90-41fe-a115-01cc8edac2d5","logging_pod":"pg-cluster-3","err":"failed to connect to `host=/controller/run user=postgres database=postgres`: server error (FATAL: the database system is starting up (SQLSTATE 57P03))"}
```

**Comment by amalic** [comment]
Do we have a solution for this? 

What about `reclaimPolicy`? When I look at the PV's I can see that the reclaim policy is set to `Delete`, would it make sense to use a strorage class with `reclaimPolicy` set to  `retain`? For the Pod restarting on another node it would have the same effect as if the server has restarted and it would continue where it left of.

**Comment by rgarcia89** [comment]
We are facing the same issue also with version 1.24.0. We have faced this now multiple times, which is super annoying since our database is of 1TB size. Destroying the replica is therefore not very attractive to us. Unfortunately I do not see any other solution currently.

**Comment by dbackeus** [comment]
Had this happen to us yesterday with CNPG 1.24 running PG17.

Hetzner had scheduled network downtime and this somehow caused our replica to end up in this broken state, failing to `pg_rewind`.

How do people get out of this situation once it's happened?

We don't seem to be able to get rid of our broken replica. We tried scaling down number of replicas from 2 to 1 but the operator fails to make progress since it seems to want to get the status of the replica but can't connect to it at all.

Our assumption was that scaling down to 1 would just kill the replica and leave the primary as is but that doesn't seem to be the case.

**Comment by dbackeus** [comment]
Just thought I'd clarify that it turned out that our specific issue here was the bug in Postgres 17.1 which could cause WAL files needed by `pg_rewind` to get deleted. It's mentioned as fixed on the 17.2 release: https://www.postgresql.org/about/news/postgresql-172-166-1510-1415-1318-and-1222-released-2965

**Comment by jeremych1000** [comment]
Following this as well. We use Rook-Ceph PVCs, and PG16 - similar issue.

**Comment by SanYattsu** [comment]
I have the same issue, but it affect only few clusters. After investigation, I identified the root cause: a missing Superuser configuration in the manifest specification. The Superuser name is postgres and the password can be any. Adding the following lines to the manifest resolved the issue:

```
spec:
  enableSuperuserAccess: true
  superuserSecret:
    name: superuser-secret
```

Manually delete the affected CNPG pod to trigger reconciliation with the updated manifest.

**Comment by Denton-L** [comment]
We seem to have run into a similar problem. We had a primary which was OOMKilled. I'm guessing that since the primary wasn't shut down cleanly, one of the WAL files wasn't uploaded properly.

When spinning up a new replica, it was reading WAL logs correctly for a while with:

```
...
restored log file "0000002B000008DD00000069" from archive
Restored WAL file
WAL restore command completed (parallel)
WAL restore command completed (parallel)
Restored WAL file
restored log file "0000002B000008DD00000069" from archive
WAL restore command completed (parallel)
```

However, we eventually encountered a missing WAL (which I confirmed in our S3 was indeed truly missing). The replica was stuck in a state of

```
Restored WAL file
started streaming WAL from primary at 8DD/6A000000 on timeline 43
could not receive data from WAL stream: ERROR: requested WAL segment 0000002B000008DD0000006A has already been removed
WAL file not found in the recovery object store
Restored WAL file
WAL restore command completed (parallel)
could not receive data from WAL stream: ERROR: requested WAL segment 0000002B000008DD0000006A has already been removed
started streaming WAL from primary at 8DD/6A000000 on timeline 43
WAL file not found in the recovery object store
waiting for WAL to become available at 8DD/6A0002A5
...
```

We worked around this by forcing a backup to happen (setting the schedule of the corresponding ScheduledBackup to the next minute), waiting for the backup to finalize and then deleting the PVC and pod and waiting for a new replica to spin up in place.

One thing that's concerning, that I'm going to be filing a separate issue about, is that the whole time, the Cluster reported "Cluster in healthy state" the whole time so we didn't realise that we had a broken replica for a while.

**Comment by pashtet04** [comment]
Hi team, we’re running into exactly the problem described here. Details below — hope this helps clarify and push toward a fix.

Operator chart version: 0.24.0
PostgreSQL image: ghcr.io/cloudnative-pg/postgresql:17
Replicas: 3
WAL PVC size: 90Gi
WAL archiving: Enabled, using barmanObjectStore (S3-compatible)

At exactly 00:00 UTC, WAL begins growing rapidly (likely daily backup or batch job).
One of the replicas consistently begins to lag in WAL replay.
WAL PVC on the primary starts filling up. When WAL PVC exceeds ~90%, the lagging replica crashes with:
```
FATAL: could not receive data from WAL stream: requested WAL segment ... has already been removed
```
CNPG does not recover the replica. It stays broken **until we manually delete the pod and its WAL PVC**.
While the broken replica exists, WAL is never cleaned up, causing infinite growth in pg_wal.

<img width="624" height="315" alt="Image" src="https://github.com/user-attachments/assets/c76ad149-420b-4745-834f-3da2069515a7" />

The cluster enters a degraded state, only 2/3 replicas are healthy.
WAL PVC usage spikes daily.
Recovery is manual (delete pod + PVC).
Alert fatigue from CNPGClusterHighReplicationLag.

Suggestions
auto-drop or fence replication slots for replicas that crash due to missing WAL.
a timeout (replicaRecoveryTimeout) to trigger eviction or re-bootstrap.
Emit a custom condition or warning in the Cluster status for this state.

[log.txt](https://github.com/user-attachments/files/21136652/log.txt)

We’re happy to share logs, metrics, or help test any patch/RCs. This is reproducible daily in our environment.

**Comment by IngLP** [comment]
+1!

**Comment by dbackeus** [comment]
@pashtet04 that Postgres image version is vague as it doesn't specify minor version. In case 17.1 was in use you may have hit this documented bug: https://github.com/cloudnative-pg/cloudnative-pg/issues/3680#issuecomment-2497920626

(if that wasn't the case I'm afraid I don't have anything more to add)

**Comment by pashtet04** [comment]
I switched image to 17.5. Two days left only
But right now I am watching for wal pvc grows again slowly
I checked the logs and occur errors

```
ERROR: replication slot \"_cnpg_pg_cluster_4\" cannot be advanced (SQLSTATE 55000)"}
This slot has never previously reserved WAL, or it has been invalidated.

SELECT pg_catalog.pg_replication_slot_advance($1, $2)
```

<img width="862" height="463" alt="Image" src="https://github.com/user-attachments/assets/06e31e34-8592-4826-a717-a5528e04312a" />

**Comment by lucasvmiguel** [comment]
First of all, I truly appreciate the work you maintainers are doing. Many people complain about free software, but few genuinely recognize the people behind it.

I started experimenting with CloudNativePG recently to see if it could be used in production. Deleting the primary pod was one of the first tests that I ran. Although there was a short downtime when deleting the primary pod, another one was elected, and the cluster was "recovered" quickly.

However, this situation of the old primary being in a permanent state of not ready was noticed in 2 completely different clouds/instances. Because of that, I would rule out anything related to the storage type.

I'm curious to see if anyone can avoid this issue, because this problem happened to me almost 100% of the time. And if that's happening to everyone, it seems a very big deal.

I'm happy to help with anything, including with code changes. The suggestions recommended by @pashtet04 are very reasonable to me (if we are sure that's not gonna cause any data loss if data consistency is on).
Should we move forward with this solution?

In case it helps, here's my setup:

cluster.yaml
```
apiVersion: postgresql.cnpg.io/v1
kind: Cluster
metadata:
  name: example-cluster
  namespace: cloud-native-pg
spec:
  imageName: ghcr.io/cloudnative-pg/postgresql:17.5
  instances: 3
  env:
  - name: TZ
    value: Europe/Amsterdam
  postgresql:
    synchronous:
      method: any
      number: 1
      dataDurability: required
    parameters:
      shared_buffers: "256MB"
      auto_explain.log_min_duration: "2s"
      pg_stat_statements.max: "10000"
      pg_stat_statements.track: all
  primaryUpdateStrategy: unsupervised
  primaryUpdateMethod: switchover
  storage:
    size: 20Gi
  bootstrap:
    initdb:
      database: app
      owner: app
      secret:
        name: cloud-native-pg-app-user
  enableSuperuserAccess: true
  superuserSecret:
    name: cloud-native-pg-app-superuser
  resources:
    requests:
      memory: "512Mi"
      cpu: "0.2"
    limits:
      memory: "1Gi"
      cpu: "1"
  monitoring:
    enablePodMonitor: true
```


--- UPDATE

After testing a few times more, I found a few pages that made me change my setup:

```yaml
...
  postgresql:
    parameters:
      ...
      wal_keep_size: "5GB" 
      max_slot_wal_keep_size: "5GB"
      hot_standby_feedback: "on"
  replicationSlots :
    highAvailability:
     enabled : true # I think this is already the default, but I just wanted to make sure anyway
```

I saw in a few places that whenever replication slots option is enabled, you don't have to change `wal_keep_size`.

That being said, adding `wal_keep_size` with a high number essentially resolved my issue of the old primary not being restored as a replica after the failover. So, I'm not seeing any more WAL segments missing. Important to mention that the cluster stays up but is not usable for around a minute


References:
* https://www.enterprisedb.com/blog/how-cloudnativepg-manages-replication-slots
* https://cloudnative-pg.io/documentation/1.26/postgresql_conf/#the-postgresql-section

**Comment by emagiz** [comment]
We also experience this issue after a switchover, the replica instance (former primary) fails to start with the following logs:
```
{
  "level": "info",
  "ts": "2025-09-19T13:11:15.650243255Z",
  "msg": "This is an old primary instance, waiting for the switchover to finish",
  "logger": "instance-manager",
  "logging_pod": "postgres-cluster-2",
  "controller": "instance-cluster",
  "controllerGroup": "postgresql.cnpg.io",
  "controllerKind": "Cluster",
  "Cluster": {
    "name": "postgres-cluster",
    "namespace": "namespace-xxxxxxx"
  },
  "namespace": "namespace-xxxxxxx",
  "name": "postgres-cluster",
  "reconcileID": "1db4ed9d-c3d0-47f4-a99a-be655d89a831",
  "currentPrimary": "postgres-cluster-1",
  "targetPrimary": "postgres-cluster-1"
}
{
  "level": "info",
  "ts": "2025-09-19T13:11:15.650251146Z",
  "msg": "Switchover completed",
  "logger": "instance-manager",
  "logging_pod": "postgres-cluster-2",
  "controller": "instance-cluster",
  "controllerGroup": "postgresql.cnpg.io",
  "controllerKind": "Cluster",
  "Cluster": {
    "name": "postgres-cluster",
    "namespace": "namespace-xxxxxxx"
  },
  "namespace": "namespace-xxxxxxx",
  "name": "postgres-cluster",
  "reconcileID": "1db4ed9d-c3d0-47f4-a99a-be655d89a831",
  "targetPrimary": "postgres-cluster-1",
  "currentPrimary": "postgres-cluster-1"
}
{
  "level": "info",
  "ts": "2025-09-19T13:11:15.650262256Z",
  "msg": "Waiting for the new primary to be available",
  "logging_pod": "postgres-cluster-2",
  "primaryConnInfo": "host=postgres-cluster-rw user=streaming_replica port=5432 sslkey=/controller/certificates/streaming_replica.key sslcert=/controller/certificates/streaming_replica.crt sslrootcert=/controller/certificates/server-ca.crt application_name=postgres-cluster-2 sslmode=verify-ca dbname=postgres connect_timeout=5"
}
{
  "level": "info",
  "ts": "2025-09-19T13:11:15.66038595Z",
  "msg": "Extracting pg_controldata information",
  "logger": "instance-manager",
  "logging_pod": "postgres-cluster-2",
  "controller": "instance-cluster",
  "controllerGroup": "postgresql.cnpg.io",
  "controllerKind": "Cluster",
  "Cluster": {
    "name": "postgres-cluster",
    "namespace": "namespace-xxxxxxx"
  },
  "namespace": "namespace-xxxxxxx",
  "name": "postgres-cluster",
  "reconcileID": "1db4ed9d-c3d0-47f4-a99a-be655d89a831",
  "reason": "before pg_rewind"
}
{
  "level": "info",
  "ts": "2025-09-19T13:11:15.661978154Z",
  "logger": "pg_controldata",
  "msg": "pg_control version number:            1700\nCatalog version number:               202406281\nDatabase system identifier:           7545450182860029973\nDatabase cluster state:               shut down\npg_control last modified:             Mon 15 Sep 2025 05:25:27 AM UTC\nLatest checkpoint location:           1/8D000028\nLatest checkpoint's REDO location:    1/8D000028\nLatest checkpoint's REDO WAL file:    00000006000000010000008D\nLatest checkpoint's TimeLineID:       6\nLatest checkpoint's PrevTimeLineID:   6\nLatest checkpoint's full_page_writes: on\nLatest checkpoint's NextXID:          0:797\nLatest checkpoint's NextOID:          16542\nLatest checkpoint's NextMultiXactId:  1\nLatest checkpoint's NextMultiOffset:  0\nLatest checkpoint's oldestXID:        730\nLatest checkpoint's oldestXID's DB:   1\nLatest checkpoint's oldestActiveXID:  0\nLatest checkpoint's oldestMultiXid:   1\nLatest checkpoint's oldestMulti's DB: 1\nLatest checkpoint's oldestCommitTsXid:0\nLatest checkpoint's newestCommitTsXid:0\nTime of latest checkpoint:            Mon 15 Sep 2025 05:25:27 AM UTC\nFake LSN counter for unlogged rels:   0/3E8\nMinimum recovery ending location:     0/0\nMin recovery ending loc's timeline:   0\nBackup start location:                0/0\nBackup end location:                  0/0\nEnd-of-backup record required:        no\nwal_level setting:                    logical\nwal_log_hints setting:                on\nmax_connections setting:              100\nmax_worker_processes setting:         32\nmax_wal_senders setting:              10\nmax_prepared_xacts setting:           0\nmax_locks_per_xact setting:           64\ntrack_commit_timestamp setting:       off\nMaximum data alignment:               8\nDatabase block size:                  8192\nBlocks per segment of large relation: 131072\nWAL block size:                       8192\nBytes per WAL segment:                16777216\nMaximum length of identifiers:        64\nMaximum columns in an index:          32\nMaximum size of a TOAST chunk:        1996\nSize of a large-object chunk:         2048\nDate/time type storage:               64-bit integers\nFloat8 argument passing:              by value\nData page checksum version:           0\nMock authentication nonce:            6ead4bf77e7c416533012bfa674af2bfb6a2dd162efa70abb690d0fb594c7174\n",
  "pipe": "stdout",
  "logging_pod": "postgres-cluster-2"
}
{
  "level": "info",
  "ts": "2025-09-19T13:11:15.665274591Z",
  "msg": "Starting up pg_rewind",
  "logger": "instance-manager",
  "logging_pod": "postgres-cluster-2",
  "controller": "instance-cluster",
  "controllerGroup": "postgresql.cnpg.io",
  "controllerKind": "Cluster",
  "Cluster": {
    "name": "postgres-cluster",
    "namespace": "namespace-xxxxxxx"
  },
  "namespace": "namespace-xxxxxxx",
  "name": "postgres-cluster",
  "reconcileID": "1db4ed9d-c3d0-47f4-a99a-be655d89a831",
  "pgdata": "/var/lib/postgresql/data/pgdata",
  "options": [
    "-P",
    "--source-server",
    "host=postgres-cluster-rw user=streaming_replica port=5432 sslkey=/controller/certificates/streaming_replica.key sslcert=/controller/certificates/streaming_replica.crt sslrootcert=/controller/certificates/server-ca.crt application_name=postgres-cluster-2 sslmode=verify-ca dbname=postgres",
    "--target-pgdata",
    "/var/lib/postgresql/data/pgdata",
    "--restore-target-wal"
  ]
}
{
  "level": "info",
  "ts": "2025-09-19T13:11:15.702816364Z",
  "logger": "pg_rewind",
  "msg": "pg_rewind: connected to server",
  "pipe": "stderr",
  "logging_pod": "postgres-cluster-2"
}
{
  "level": "info",
  "ts": "2025-09-19T13:11:15.709403836Z",
  "logger": "pg_rewind",
  "msg": "pg_rewind: servers diverged at WAL location 1/87000000 on timeline 6",
  "pipe": "stderr",
  "logging_pod": "postgres-cluster-2"
}
{
  "level": "info",
  "ts": "2025-09-19T13:11:15.709608789Z",
  "logger": "pg_rewind",
  "msg": "pg_rewind: error: could not read file \"/var/lib/postgresql/data/pgdata/pg_wal/000000060000000100000086\": read 0 of 8192",
  "pipe": "stderr",
  "logging_pod": "postgres-cluster-2"
}
{
  "level": "info",
  "ts": "2025-09-19T13:11:15.709660472Z",
  "logger": "pg_rewind",
  "msg": "pg_rewind: error: could not find previous WAL record at 1/860000F8",
  "pipe": "stderr",
  "logging_pod": "postgres-cluster-2"
}
{
  "level": "error",
  "ts": "2025-09-19T13:11:15.710883013Z",
  "msg": "Failed to execute pg_rewind",
  "logger": "instance-manager",
  "logging_pod": "postgres-cluster-2",
  "controller": "instance-cluster",
  "controllerGroup": "postgresql.cnpg.io",
  "controllerKind": "Cluster",
  "Cluster": {
    "name": "postgres-cluster",
    "namespace": "namespace-xxxxxxx"
  },
  "namespace": "namespace-xxxxxxx",
  "name": "postgres-cluster",
  "reconcileID": "1db4ed9d-c3d0-47f4-a99a-be655d89a831",
  "options": [
    "-P",
    "--source-server",
    "host=postgres-cluster-rw user=streaming_replica port=5432 sslkey=/controller/certificates/streaming_replica.key sslcert=/controller/certificates/streaming_replica.crt sslrootcert=/controller/certificates/server-ca.crt application_name=postgres-cluster-2 sslmode=verify-ca dbname=postgres",
    "--target-pgdata",
    "/var/lib/postgresql/data/pgdata",
    "--restore-target-wal"
  ],
  "error": "exit status 1",
  "stacktrace": "github.com/cloudnative-pg/machinery/pkg/log.(*logger).Error\n\tpkg/mod/github.com/cloudnative-pg/machinery@v0.0.0-20241219102532-2807bc88310d/pkg/log/log.go:125\ngithub.com/cloudnative-pg/cloudnative-pg/pkg/management/postgres.(*Instance).Rewind\n\tpkg/management/postgres/instance.go:1040\ngithub.com/cloudnative-pg/cloudnative-pg/internal/management/controller.(*InstanceReconciler).verifyPgDataCoherenceForPrimary\n\tinternal/management/controller/instance_startup.go:271\ngithub.com/cloudnative-pg/cloudnative-pg/internal/management/controller.(*InstanceReconciler).initialize\n\tinternal/management/controller/instance_controller.go:408\ngithub.com/cloudnative-pg/cloudnative-pg/internal/management/controller.(*InstanceReconciler).Reconcile\n\tinternal/management/controller/instance_controller.go:164\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).Reconcile\n\tpkg/mod/sigs.k8s.io/controller-runtime@v0.19.3/pkg/internal/controller/controller.go:116\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).reconcileHandler\n\tpkg/mod/sigs.k8s.io/controller-runtime@v0.19.3/pkg/internal/controller/controller.go:303\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).processNextWorkItem\n\tpkg/mod/sigs.k8s.io/controller-runtime@v0.19.3/pkg/internal/controller/controller.go:263\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).Start.func2.2\n\tpkg/mod/sigs.k8s.io/controller-runtime@v0.19.3/pkg/internal/controller/controller.go:224"
}
{
  "level": "error",
  "ts": "2025-09-19T13:11:15.711099987Z",
  "msg": "Reconciler error",
  "logger": "instance-manager",
  "logging_pod": "postgres-cluster-2",
  "controller": "instance-cluster",
  "controllerGroup": "postgresql.cnpg.io",
  "controllerKind": "Cluster",
  "Cluster": {
    "name": "postgres-cluster",
    "namespace": "namespace-xxxxxxx"
  },
  "namespace": "namespace-xxxxxxx",
  "name": "postgres-cluster",
  "reconcileID": "1db4ed9d-c3d0-47f4-a99a-be655d89a831",
  "error": "while exucuting pg_rewind: error executing pg_rewind: exit status 1",
  "stacktrace": "sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).reconcileHandler\n\tpkg/mod/sigs.k8s.io/controller-runtime@v0.19.3/pkg/internal/controller/controller.go:316\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).processNextWorkItem\n\tpkg/mod/sigs.k8s.io/controller-runtime@v0.19.3/pkg/internal/controller/controller.go:263\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller[...]).Start.func2.2\n\tpkg/mod/sigs.k8s.io/controller-runtime@v0.19.3/pkg/internal/controller/controller.go:224"
}
{
  "level": "info",
  "ts": "2025-09-19T13:11:19.426454382Z",
  "msg": "Unable to collect metrics",
  "logging_pod": "postgres-cluster-2",
  "error": "failed to connect to `user=postgres database=postgres`: /controller/run/.s.PGSQL.5432 (/controller/run): dial error: dial unix /controller/run/.s.PGSQL.5432: connect: no such file or directory"
}
{
  "level": "info",
  "ts": "2025-09-19T13:11:35.735393793Z",
  "msg": "Unable to collect metrics",
  "logging_pod": "postgres-cluster-2",
  "error": "failed to connect to `user=postgres database=postgres`: /controller/run/.s.PGSQL.5432 (/controller/run): dial error: dial unix /controller/run/.s.PGSQL.5432: connect: no such file or directory"
}
```
It seems to complain about reading the WAL, I can see that the WAL size it complains about is 0KB which seems corrupted can explain why it is failing to start. The only way to get out of this situation is to delete the replica and its PVCs, but the question here how can I protect that it does not happen again (given the condition that the primary pod gets terminated ungracefully) 

Operator version: 1.25
Postgres version: 17.2

**Comment by dragoangel** [comment]
> I have the same issue, but it affect only few clusters. After investigation, I identified the root cause: a missing Superuser configuration in the manifest specification. The Superuser name is postgres and the password can be any. Adding the following lines to the manifest resolved the issue:
> 
> ```
> spec:
>   enableSuperuserAccess: true
>   superuserSecret:
>     name: superuser-secret
> ```
> 
> Manually delete the affected CNPG pod to trigger reconciliation with the updated manifest.

Well this is not correct, there no need to set password as pg_rewind and physical replication between master & standy (read replicas of 1 cluster) done via tls keys out of the box.

**Comment by awoimbee** [comment]
Well I just had a cluster die because of this, all pods were dead so in the end I had to recover from a backup, which didn't work because of errors `could not locate required checkpoint record at 1C/CD000080`.
Each time I tried an earlier backup and a different `recoveryTarget`, but got the same error, with a different LSN:  `1C/CD000080` `1D/C7000060` `1E/8000060` `1E/C000060` `1E/2F000060` ...

**Comment by 5h4k4r** [comment]
I'm experiencing the same error. You mentioned that deleting the PVC resolves the issue, but what happens to the data? In my case, the cluster failed over to another pod, and now the old master pod is failing:

`{"level":"debug","ts":"2025-10-30T10:25:26.016355877Z","msg":"Instance status probe failing","caller":"pkg/mod/github.com/cloudnative-pg/machinery@v0.3.1/pkg/log/log.go:178","logging_pod":"cn-postgres-cluster-1","err":"failed to connect to `user=postgres database=postgres`: /controller/run/.s.PGSQL.5432 (/controller/run): dial error: dial unix /controller/run/.s.PGSQL.5432: connect: no such file or directory"}
`

 I'm concerned about potential data loss if I delete the PVC of the old master.

How can I ensure that the new master is fully synchronized with the old one?

**Comment by archcore-dev** [comment]
> I'm experiencing the same error. You mentioned that deleting the PVC resolves the issue, but what happens to the data? In my case, the cluster failed over to another pod, and now the old master pod is failing:
> 
> `{"level":"debug","ts":"2025-10-30T10:25:26.016355877Z","msg":"Instance status probe failing","caller":"pkg/mod/github.com/cloudnative-pg/machinery@v0.3.1/pkg/log/log.go:178","logging_pod":"cn-postgres-cluster-1","err":"failed to connect to `user=postgres database=postgres`: /controller/run/.s.PGSQL.5432 (/controller/run): dial error: dial unix /controller/run/.s.PGSQL.5432: connect: no such file or directory"} `
> 
> I'm concerned about potential data loss if I delete the PVC of the old master.
> 
> How can I ensure that the new master is fully synchronized with the old one?

Same problem here I can't delete pvc how to solve it

**Comment by darkpixel** [comment]
Stale bots are dumb.
It's not stale because it's waiting on users, it's stale because the issue hasn't been fixed yet. ;)

**Comment by Brice187** [comment]
still happens on 

```
helmCharts:
  - name: cloudnative-pg
    version: "0.27.1"

image: ghcr/cloudnative-pg/postgresql:18-standard-trixie
```

**Comment by davidpanic** [comment]
[Stale bots suck](https://nostalebots.xyz/). This is still an issue.

**Comment by amalic** [comment]
This issue needs a solution, otherwise CNPG cannot be used in production!
